### ISSUE TITLE
[Feature Discussion] Nested Reflected Types

### ISSUE THREAD

**Issue by asodeur** [numba/numba#3546]
Currently it is not possible to nest reflected types like `list` or `set` (see tests marked with `@expect_reflection_failure`) because the Python refcount of the inner types is not managed.

A naive imlementation would simply add managing the refcount of inner types by increfing when a reference is obtained and decrefing when the `MemInfo` refcount goes to zero (see PR coming soon). However, there are requirements that will probably make things more complicated. This is an attempt to gather those, starting with what I am already aware of.

1. GIL: Touching the refcount of a Python object requires the GIL. Decrefing the Python object immediately when the `meminfo` refcount goes to zero might result in aquiring/releasing the GIL rather often. It might be enough to decref the Python objects before leaving `nopython` code but that is tricky when multi-threading.
2. Littering on `MemInfo`: Naively tracking the owner objects can be implemented by adding members to `MemInfo`. Memory impact needs to be understood.
3. There seem to be use-cases where Numba is used without a Python runtime [dynd-python](https://github.com/libdynd/dynd-python) which will probably create extra constraints on the implementation.
4. Even with Python runtime  there is a need for objects created in C (see #2810). Creating arrays of reflected types in C might be harder as from the Python-side where full (Numba-)type information is available

There are probably a lot more issues to consider. Please add.

**Comment by seibert** [comment]
Thanks for bringing up this as a discussion topic!  I can chime in with a summary of the core developer thinking on this topic.

Reflection was initially added as a mechanism to "transparently" (more on that) handle list arguments to Numba-compiled functions.  On one hand, this has made it possible to apply Numba to more use cases involving lists, but on the other hand, it has a number of shortcomings which are made even bigger by nested data structures:

* Unboxing of Python lists is costly because we must walk the entire list and individually unbox each element.  Creating a parallel list representation is much slower than unboxing other types, like scalars, NumPy arrays, and strings.
* List reflection creates an inconsistent state during execution of the nopython function, where the list in Python and the translated list in nopython mode can have different contents.  The reflection step at the end of the function call patches things up so things look consistent again, but in the presence of multiple threads, there is no consistent view.  (It is fair to point out that without some locking strategy, any modification of lists between different threads is asking for a race condition.  That problem is potentially fixable, but not in the current situation with list reflection.)
* Reflecting the list contents back to the original Python list at the end of the function is also slow, and becomes significantly more complex for all the reasons mentioned in the original issue description.

For these reasons, we are strongly leaning toward declaring that the list reflection technique is an anti-pattern for Numba boxing/unboxing, and instead we'd like to shift to the following approach for container types:
* When Numba supports a Python container (like list), we support creation of a statically typed equivalent (like "list of list of float64") in a nopython-compatible data structure.
* We will supply a Python wrapper around this nopython data structure allowing it to be used from Python as if it were a normal Python container.  This interpreter access may be (in fact, likely will be) slower than the native Python container.  We would call this type a "Numba X", ex: Numba list, Numba dict, etc.
* If an instance of the container is created inside nopython mode, it will be mutable (if that makes sense) and if returned to Python, will be return as a Numba list/dict/etc.  Numba will not translate it back to a native Python list/dict/etc.
* If the native Python container is passed as an argument to a nopython mode function, Numba will unbox it into the equivalent Numba container.  In these situations, the object will be treated as immutable and attempts to modify it will raise compilation errors.
* Users who want to mutate containers passed in as arguments (and avoid the unboxing overhead) will be directed toward creating instances of the Numba container outside nopython mode.  This makes the translation between Numba and Python containers explicit (and hopefully infrequent).
* Explicit translation between Python and Numba containers will be easy to do.

Now, this is easy to say, but harder to do.  We are not going to rip out the current list support, and will likely first try out this approach with Numba dictionary support in the near future.  If that turns out to be successful (and the restrictions are not too onerous), we'll go back and redo lists in a similar way.

tldr; Regarding nested reflected types, our current preference is "Don't do it!  There is a better way."

**Comment by asodeur** [comment]
Stan,

the approach outlined above indeed seems to avoid a lot of trouble while not sacrificing too much functionailty for list/dict/etc.
 
We are currently using custom reflected types to keep a link back to the Python object, ie our Numba-type is a 'simplified' version of the full Python object that never gets mutated from nopython mode. Hence we never actually reflect changes back from nopython but we need the pairing of Python/nopython objects. 

I've seen #3282 and it seems to me that this could provide this functionality. Is there something to look at already? 

Andreas
